### ISSUE TITLE
PEP 681 - Wrong init signature reported when adding fields through a decorator

### ISSUE THREAD

**Issue by tjoloi** [microsoft/pyright#5633]
**Describe the bug**
When using the `@dataclass_transform()` decorator from PEP 681 on a decorator function that modify the decorated class' fields, pyright doesn't include these new fields when building the __init__ function.

Consider the following code:
```python
from typing import dataclass_transform, cast
from dataclasses import make_dataclass

@dataclass_transform()
def decorator(cls):
    cls.__annotations__['custom_field'] = float
    fields = list(cls.__annotations__.items())
    dataCls = make_dataclass(cls.__name__, fields, bases=(cls,))
    return dataCls

@decorator
class my_dataclass:
    attr1: int
    attr2: str

c = my_dataclass(1, 'a', custom_field = 1.0)
```
The idea is to use the default `make_dataclass` provided by python but add a common field that doesn't need to be specified in the decorated class. When doing so, `make_dataclass` correctly adds a `custom_field` parameter to the generated init function.

Unfortunately, when using `reveal_type(my_dataclass.__init__)` I get this:
```python
(function) def __init__(
    self: my_dataclass,
    attr1: int,
    attr2: str
```
and pylance throws the following issue in the code: `No parameter named "custom_field"`

Before the introduction of `dataclass_transform` this code used to work in most cases but I've been exploring it because of an issue when casting to `my_dataclass`.
`typing.cast(my_dataclass, c)` returns a `type` object instead of the casted type (`Type[type]` instead of `Type[my_dataclass]`)

Adding the following code at the end of the script should highlight the second issue when removing the `dataclass_transform` decorator:
```python
c = cast(my_dataclass, c)
print(c.attr1)
```

In both cases, the problem is only with the static type checking, the code still runs without any issue.

**VS Code extension or command-line**
PyLance v2023.7.40

**Comment by erictraut** [comment]
Pyright is working correctly here. PEP 681 does not provide any provision for the dynamic addition of custom fields like you're doing in this code sample.

It wouldn't be appropriate for PEP 681 to work the way you suggest. When applying a decorator (regardless of whether it is decorated with `dataclass_transform`), the evaluated type must be based on the _signature_ of the decorator, not the _implementation_ of the decorator. The implementation is never considered by a static type checker when evaluating the type of a call expression (which is what a decorator is). 

Consider if your decorator were declared in a stub file. The implementation would be replaced with `...`.

```python
@dataclass_transform()
def decorator(cls):
    ...
```

There's no information in the `decorator` signature that would suggest that there are additional custom fields present.
